### PR TITLE
Implement PageView using SliverLayoutBuilder, Deprecate RenderSliverFillViewport

### LINKED ISSUE #23873: PageView with viewportFraction < 1.0 - GestureDetector is not working properly
Hey, looks like there is a bug preventing gesture detection on right side of the widget inside page view with only one element and viewport fraction smaller than 1.0.

This is how my widget tree looks like:
```
column
  - page view
    - gesture detector
      - green container
    - gesture detector
      - blue container
  - page view
    - gesture detector
      - red container
```

screenshot from 2018-11-02 16-14-59

Top gesture detectors (green&blue widgets) work fine, I can click anywhere and I can see debug logs in console.
Bottom gesture detector (red widget) works fine only for left half of the widget. When I click on the right half then onTap callback is not fired. Looks like something is preventing gesture detector from getting input.

flutter doctor output:
```
[✓] Flutter (Channel beta, v0.9.4, on Linux, locale en_US.UTF-8)
[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
[✓] Android Studio (version 3.1)
[!] IntelliJ IDEA Community Edition (version 2018.2)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[✓] Connected devices (1 available)
```

Source code to reproduce bug:
```
import 'package:flutter/material.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Flutter Demo',
      theme: new ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Column(children: <Widget>[
        Expanded(
          child: ColorPages(
            colors: <Color>[
              Colors.green,
              Colors.blue,
            ],
          ),
        ),
        Expanded(
          child: ColorPages(
            colors: <Color>[
              Colors.red,
            ],
          ),
        ),
      ]),
    );
  }
}

class ColorPages extends StatefulWidget {
  final List<Color> colors;

  const ColorPages({Key key, this.colors}) : super(key: key);

  @override
  ColorPagesState createState() {
    return new ColorPagesState();
  }
}

class ColorPagesState extends State<ColorPages> {
  PageController _pageController;

  @override
  void initState() {
    super.initState();

    _pageController = new PageController(
      viewportFraction: 0.5,
    );
  }

  @override
  Widget build(BuildContext context) {
    return PageView.builder(
      controller: _pageController,
      itemBuilder: (context, index) {
        return GestureDetector(
          onTap: () {
            print(DateTime.now().toIso8601String() + " " + widget.colors[index].toString());
          },
          child: Container(
            color: widget.colors[index],
          ),
        );
      },
      itemCount: widget.colors.length,
    );
  }
}
```

### LINKED ISSUE #27744: PageView doesn't do hit testing right with padding (SliverFillViewport hit test issue?)
`InkWell` will not work if `_file.length` is 1;
but it will work if length is greater than or equal to 2; 
```dart
  Container(
    height: 200,
    child: PageView.builder(
      itemCount: _file.length,
      controller: PageController(viewportFraction: 0.4),
      itemBuilder: (context, index) {
        return Card(
          child: InkWell(
            onTap: () => _picImage(),
            child: Center(
              child: Icon(
                Icons.add,
                size: 40,
              ),
            ),
          ),
        );
      },
    ),
  )
```

```
flutter doctor -v

[√] Flutter (Channel beta, v1.0.0, on Microsoft Windows [Version 10.0.17758.1], locale en-AS)
    • Flutter version 1.0.0 at C:\flutter
    • Framework revision 5391447fae (2 months ago), 2018-11-29 19:41:26 -0800
    • Engine revision 7375a0f414
    • Dart version 2.1.0 (build 2.1.0-dev.9.4 f9ebf21297)

[√] Android toolchain - develop for Android devices (Android SDK 28.0.3)
    • Android SDK at C:\Users\Suraj\AppData\Local\Android\sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)
    • All Android licenses accepted.

[√] Android Studio (version 3.3)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 32.0.1
    • Dart plugin version 182.5215
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)

[!] IntelliJ IDEA Ultimate Edition (version 2018.2)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA 2018.2.3
    X Flutter plugin not installed; this adds Flutter specific functionality.
    X Dart plugin not installed; this adds Dart specific functionality.
    • For information about installing plugins, see
      https://flutter.io/intellij-setup/#installing-the-plugins
```

PS: Using Android Studio Not Intellij for Flutter

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
## Description

- Deprecate `RenderSliverFillViewport`
- Change `PageView`'s `build` method to use `SliverLayoutBuilder` instead.

## Breaking Changes
- `RenderSliverFillViewport` is now marked as `deprecated`.

## Related Issues

Closes https://github.com/flutter/flutter/issues/23873
Closes https://github.com/flutter/flutter/issues/27744

## Tests

I added the following tests:

All visible pages are able to receive touch events.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/sliver.dart [hunk 1]
```diff
@@ -1030,6 +1028,7 @@ class SliverGrid extends SliverMultiBoxAdaptorWidget {
 ///    the main axis extent of each item.
 ///  * [SliverList], which does not require its children to have the same
 ///    extent in the main axis.
+@Deprecated('Use SliverLayoutBuilder instead.')
 class SliverFillViewport extends SliverMultiBoxAdaptorWidget {
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/sliver.dart]: Can we keep this widget and just implement it in terms of SliverLayoutBuilder and SliverFixedExtendList? That would limit the scope of the API breakage and we'd only remove the RenderSliverFillViewport, I think?

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/sliver.dart]: Aye

File: packages/flutter/test/widgets/page_view_test.dart [hunk 2]
```diff
@@ -556,13 +556,59 @@ void main() {
 
     await tester.pumpWidget(build(controller));
 
-    expect(tester.getTopLeft(find.text('Alabama')), const Offset(-100.0, 0.0));
-    expect(tester.getBottomRight(find.text('Alabama')), const Offset(900.0, 600.0));
+    expect(tester.getTopLeft(find.text('Alabama')), Offset.zero);
```
Comment (chunhtai) [inline packages/flutter/test/widgets/page_view_test.dart]: why this change?

Comment (chunhtai) [reply packages/flutter/test/widgets/page_view_test.dart]: it looks like we changed the behavior when viewportFraction > 1
The page start at the very left edge of the page instead of after extra space.

**Review by chunhtai — COMMENTED** [review]
overall looks good, but one concern about the behavior when viewportFraction > 1.0

File: packages/flutter/test/widgets/page_view_test.dart (hunk 2 above)
Comment (LongCatIsLooong) [reply packages/flutter/test/widgets/page_view_test.dart]: Yeah it does change the behavior when `viewportFraction > 1` and breaks the API (although I doubt it would really affect anyone, and starting from 0 makes more sense to me) . Context: https://github.com/flutter/flutter/pull/8539

File: packages/flutter/lib/src/widgets/sliver.dart [hunk 3]
```diff
@@ -1047,15 +1052,31 @@ class SliverFillViewport extends SliverMultiBoxAdaptorWidget {
...
   @override
-  void updateRenderObject(BuildContext context, RenderSliverFillViewport renderObject) {
-    renderObject.viewportFraction = viewportFraction;
+  Widget build(BuildContext context) {
+    return SliverLayoutBuilder(
+      builder: (BuildContext context, SliverConstraints constraints) {
+        final double fixedExtent = constraints.viewportMainAxisExtent * viewportFraction;
+        final double padding = math.max(0, constraints.viewportMainAxisExtent - fixedExtent) / 2;
+
+        final bool isHorizontal = constraints.axisDirection == AxisDirection.left
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/sliver.dart]: Can you just use constraints.axis?

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/sliver.dart]: Also, with enums instead of checking individual values use a switch/case. That way the analyzer tells us that we need to update this code if we ever add new enum values.

File: packages/flutter/test/widgets/page_view_test.dart (hunk 2 above)
Comment (goderbauer) [reply packages/flutter/test/widgets/page_view_test.dart]: What's the motivation for changing this?

Please add a description of the changed behavior to the PR description.

File: packages/flutter/test/widgets/page_view_test.dart [hunk 4]
```diff
@@ -556,13 +556,59 @@ void main() {
...
 
     controller.jumpToPage(10);
     await tester.pump();
 
-    expect(tester.getTopLeft(find.text('Hawaii')), const Offset(-100.0, 0.0));
+    expect(tester.getTopLeft(find.text('Hawaii')), Offset.zero);
+  });
+
+  testWidgets('All visible pages are able to receive touch events',
+    (WidgetTester tester) async {
```
Comment (goderbauer) [inline packages/flutter/test/widgets/page_view_test.dart]: nit: either place this at the end of the previous line (preferred) or indent this more.

File: packages/flutter/test/widgets/page_view_test.dart [hunk 5]
```diff
@@ -556,13 +556,59 @@ void main() {
...
+    Widget build() {
+      return Directionality(
+        textDirection: TextDirection.ltr,
+        child: PageView.builder(
+          controller: controller,
+          itemCount: 20,
+          itemBuilder: (BuildContext context, int index) {
+            return GestureDetector(
+              onTap: () => tappedIndex = index,
+              child: SizedBox.expand(child: Text(index.toString())),
```
Comment (goderbauer) [inline packages/flutter/test/widgets/page_view_test.dart]:
```suggestion
              child: SizedBox.expand(child: Text('$index')),
```

File: packages/flutter/test/widgets/page_view_test.dart [hunk 6]
```diff
@@ -556,13 +556,59 @@ void main() {
...
 
-    expect(tester.getTopLeft(find.text('Hawaii')), const Offset(-100.0, 0.0));
+    expect(tester.getTopLeft(find.text('Hawaii')), Offset.zero);
+  });
+
+  testWidgets('All visible pages are able to receive touch events',
+    (WidgetTester tester) async {
+    final PageController controller = PageController(viewportFraction: 1/4, initialPage: 0);
+    int tappedIndex;
+    Iterable<int> visiblePages;
```
Comment (goderbauer) [inline packages/flutter/test/widgets/page_view_test.dart]: Declare this were the first value is assigned to it?

File: packages/flutter/test/widgets/page_view_test.dart (hunk 5 above)
Comment (goderbauer) [reply packages/flutter/test/widgets/page_view_test.dart]: Also below - I think that improves readability.

File: packages/flutter/test/widgets/page_view_test.dart [hunk 7]
```diff
@@ -556,13 +556,59 @@ void main() {
...
+
+    visiblePages = const <int> [0, 1, 2];
+    await tester.pumpWidget(build());
+
+    // The first 3 items should be visible and tappable.
+    for (int index in visiblePages) {
+      expect(find.text(index.toString()), findsOneWidget);
+      // The center of page 2's x-coordinate is 800, so we have to manually
+      // offset it a bit to make sure the tap lands within the screen.
+      final Offset center = tester.getCenter(find.text(index.toString())).translate(-3, 0);
```
Comment (goderbauer) [inline packages/flutter/test/widgets/page_view_test.dart]: You could also just subtract a small offset, which may make the intend here clearer:

```suggestion
      final Offset center = tester.getCenter(find.text('$index')) - const Offset(3, 0);
```

**Review by goderbauer — COMMENTED** [review]
Can you update the PR's description? Please also include all changes in behavior there.

You'll also have to follow https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes for both of the breaking changes.

File: packages/flutter/test/widgets/page_view_test.dart (hunk 2 above)
Comment (chunhtai) [reply packages/flutter/test/widgets/page_view_test.dart]: I actually think the original behavior make more sense to me. I would expect the viewport is center align with the page, that is why you have the padding when viewportfraction<1. If the same thing apply when viewportfraction>1, the original behavior is expected. and i am not quite sure how https://github.com/flutter/flutter/pull/8539 related to this behavior. Can you elaborate?

Comment (LongCatIsLooong) [reply packages/flutter/test/widgets/page_view_test.dart]: Ah I thought a weird `viewportFraction` like 1.3 would mess up the snapping behavior but that's not true. Thanks for pointing out. But `SliverPadding`s can't go negative. I'll find an alternative for it.

File: packages/flutter/lib/src/widgets/page_view.dart [hunk 8]
```diff
@@ -364,22 +364,24 @@ class _PagePosition extends ScrollPositionWithSingleContext implements PageMetri
       return;
     final double oldPage = page;
     _viewportFraction = value;
+
     if (oldPage != null)
-      forcePixels(getPixelsFromPage(oldPage));
+      // This part should only be reacheable when it will definitely trigger a rebuild.
+      correctPixels(getPixelsFromPage(oldPage));
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/page_view.dart]: this feel dangerous to me. That is true when you change this property in didupdatewidget. If you are certain that is the case for the future as well, you should add an assert to verify that.

File: packages/flutter/lib/src/widgets/page_view.dart [hunk 9]
```diff
@@ -364,22 +364,24 @@ class _PagePosition extends ScrollPositionWithSingleContext implements PageMetri
       return;
     final double oldPage = page;
     _viewportFraction = value;
+
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/page_view.dart]: nit, extra line

File: packages/flutter/test/widgets/page_view_test.dart [hunk 10]
```diff
@@ -556,13 +556,95 @@ void main() {
 
     await tester.pumpWidget(build(controller));
 
-    expect(tester.getTopLeft(find.text('Alabama')), const Offset(-100.0, 0.0));
+    expect(tester.getTopLeft(find.text('Alabama')), const Offset(-100, 0));
```
Comment (chunhtai) [inline packages/flutter/test/widgets/page_view_test.dart]: let's revert this change

File: packages/flutter/lib/src/widgets/page_view.dart (hunk 8 above)
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/page_view.dart]: I am not sure if this is better but how about
```
if (SchedulerBinding.instance.schedulerPhase == SchedulerPhase.persistentCallbacks) {
     correctPixels(getPixelsFromPage(oldPage));
    } else {
      forcePixels(getPixelsFromPage(oldPage));
    }

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/page_view.dart]: This class itself is private, and the setter is only used in `PageController.attach`. I'm not sure if we should call `notifyListeners` in `ScrollController.attach`, I think it should be up to the caller?

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/page_view.dart]: I am more worry about we change the scroll offset but not markneedbuild, and yes this should be fine because this is private class and as of now it only called during the build stage. However, it is possible this might change. If you think that is unlikely, we should still add an assert here to guard it.

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/page_view.dart]: Yeah let me get rid of the setter altogether. The implementation should probably live in `PageController.attach` which has more context.

**Comment by LongCatIsLooong** [comment]
@goderbauer It no longer changes the behavior when `viewFraction > 1` I believe. Should I post the breakage to flutter-announcement now or there are further changes that need to be made?

File: packages/flutter/lib/src/widgets/page_view.dart [hunk 11]
```diff
@@ -252,7 +252,14 @@ class PageController extends ScrollController {
   void attach(ScrollPosition position) {
     super.attach(position);
     final _PagePosition pagePosition = position;
-    pagePosition.viewportFraction = viewportFraction;
+
+    if (pagePosition.viewportFraction == viewportFraction)
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/page_view.dart]: from `ScrollPosition.forcePixel` documentation:
>   /// This should not be called during layout (e.g. when setting the initial
  /// scroll offset). Consider [correctPixels] if you find you need to adjust
  /// the position during layout.

Also it doesn't sound right to notify the listeners in `ScrollController.attach`.

[the remark this answers is not in the document]
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/page_view.dart]: Why move it out of the setter though?

[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/page_view.dart]: IMO since using either `forcePixel` or `correctPixel` needs to be decided on a case-by-case basis, it feels a bit error-prone to have a `viewportFraction` setter which sounds like a silver bullet for changing `viewportFraction`.
Also `attach` is the only caller of the setter right now, so I got rid of the setter and moved the implementation to `attach`.

File: packages/flutter/test/widgets/page_view_test.dart [hunk 12]
```diff
@@ -565,6 +565,88 @@ void main() {
...
+      await tester.pumpWidget(build(newController));
+      newController.jumpToPage(10);
+      await tester.pump();
+
+      expect(tester.getTopLeft(find.text('Hawaii')), const Offset(-(4 - 1) * 800 / 2, 0));
+  });
+
+  testWidgets(
+    'All visible pages are able to receive touch events',
+    (WidgetTester tester) async {
```
Comment (Piinks) [inline packages/flutter/test/widgets/page_view_test.dart]: Same nit as above here regarding formatting.

File: packages/flutter/lib/src/widgets/page_view.dart [hunk 13]
```diff
@@ -252,7 +252,14 @@ class PageController extends ScrollController {
...
+      return;
+    final double oldPage = pagePosition.page;
+    pagePosition._viewportFraction = viewportFraction;
+    if (oldPage != null) {
+      pagePosition.correctPixels(pagePosition.getPixelsFromPage(oldPage));
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/page_view.dart]: according to documentation of correctPixels, this function should only be called during layout phase
can you add an assert something like
assert(SchedulerBinding.instance.schedulerPhase == SchedulerPhase.persistentCallbacks, 'attach can only be called during SchedulerPhase.persistentCallbacks')

If this assumption no longer held, would be better to update the documentation of correctPixels to include this case

File: packages/flutter/test/widgets/page_view_test.dart (hunk 12 above)
Comment (LongCatIsLooong) [reply packages/flutter/test/widgets/page_view_test.dart]: I double checked the indentation, line 607 does have 2 more spaces than 606, it's just `(` takes much less space than a regular character.

File: packages/flutter/lib/src/widgets/page_view.dart (hunk 11 above)
[the remark this answers is not in the document]
Comment (Hixie) [reply packages/flutter/lib/src/widgets/page_view.dart]: this is build, not layout. should be fine to call during build.

File: packages/flutter/lib/src/widgets/page_view.dart (hunk 13 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/page_view.dart]: should really only be called when debugDoingLayout is true for the viewport in question

File: packages/flutter/lib/src/widgets/gesture_detector.dart [hunk 14]
```diff
@@ -930,8 +930,7 @@ class RawGestureDetectorState extends State<RawGestureDetector> {
     }
   }
 
-  /// This method can be called outside of the build phase to filter the list of
-  /// available semantic actions.
+  /// This method can be called to filter the list of available semantic actions.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/gesture_detector.dart]: This should document when it is valid to call this method (e.g. what the assert is asserting).

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/page_view.dart [hunk 15]
```diff
@@ -368,18 +368,24 @@ class _PagePosition extends ScrollPositionWithSingleContext implements PageMetri
...
     if ((actual - round).abs() < precisionErrorTolerance) {
       return round;
     }
     return actual;
   }
 
-  double getPixelsFromPage(double page) {
-    return page * viewportDimension * viewportFraction;
-  }
+  double getPixelsFromPage(double page) => page * viewportDimension * viewportFraction + _initialPageOffset;
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/page_view.dart]: I think it has to be a good old function if the line is too long

**Review by chunhtai — APPROVED** [review]
LGTM, with some nit

**Comment by tvolkert** [comment]
It looks like this regressed the "stock_build_iteration" microbenchmark by about 2%.

Screen Shot 2019-08-22 at 8 51 51 AM

@LongCatIsLooong can you investigate?  https://github.com/flutter/flutter/issues/39060

File: packages/flutter/lib/src/rendering/sliver_fill.dart [hunk 16]
```diff
@@ -27,6 +27,7 @@ import 'sliver_multi_box_adaptor.dart';
 ///  * [RenderSliverFixedExtentList], which has a configurable [itemExtent].
 ///  * [RenderSliverList], which does not require its children to have the same
 ///    extent in the main axis.
+@Deprecated('Use SliverLayoutBuilder instead.')
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/sliver_fill.dart]: Looking into this further, I'm not sure this deprecation makes sense. If someone is using RenderSliverFillViewport directly, then they're not using widgets, which means that using SliverLayoutBuilder isn't an option.

We should probably remove this render object if we're not supporting it any more (which I presume we're not, since we didn't try to fix the hit test bug in this render object).

cc @goderbauer @Piinks

Comment (chunhtai) [reply packages/flutter/lib/src/rendering/sliver_fill.dart]: we move the padding from rendering layer to widget layer. I can't think of a way to achieve the same behavior by using our current rendering library. I agree we should remove this given we have deprecated warning for a while.

Comment (goderbauer) [reply packages/flutter/lib/src/rendering/sliver_fill.dart]: Yes, this is no longer used and should just be removed: https://github.com/flutter/flutter/issues/44612
